pkg/operator/mon/health.go

```diff
@@ -143,9 +143,64 @@ func (c *Cluster) checkHealth() error {
 		return nil
 	}
 
+	// check if there are more than two mons running on the same node, failover one mon in that case
+	done, err := c.checkMonsOnSameNode()
+	if done || err != nil {
+		return err
+	}
+
+	done, err = c.checkMonsOnValidNodes()
+	if done || err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (c *Cluster) checkMonsOnSameNode() (bool, error) {
+	nodesUsed := map[string]struct{}{}
+	for name, node := range c.mapping.Node {
+		// when the node is already in the list we have more than one mon on that node
+		if _, ok := nodesUsed[node.Name]; ok {
+			// get list of available nodes for mons
+			availableNodes, _, err := c.getAvailableMonNodes()
+			if err != nil {
+				return true, fmt.Errorf("failed to get available mon nodes. %+v", err)
+			}
+			// if there are enough nodes for mons failover the mon too much from the node
+			if len(availableNodes) > 0 {
+				logger.Infof("rebalance: enough nodes available %d to failover mon %s", len(availableNodes), name)
+				c.failMon(len(c.clusterInfo.Monitors), name)
+			} else {
+				logger.Warningf("rebalance: not enough nodes available to failover mon %s", name)
+			}
+
+			// deal with one mon too much on a node at a time
+			return true, nil
+		}
+		nodesUsed[node.Name] = struct{}{}
+	}
+	return false, nil
+}
+
+func (c *Cluster) checkMonsOnValidNodes() (bool, error) {
```

Comment:
is this function necessary? `getAvailableMonNodes()`, which is called by `checkMonsOnSameNode()`, already ensures that the available node it returns is `valid`.

Reply:
This function/check `checkMonsOnValidNodes()` goes over all nodes used and checks if they are still valid (placement setting update) and if not fail the mon over to another "new" available node.

@travisn Additionally if the `c.placement` is updated during runtime this could potentially allow mons to be moved/failovered one by one to "new" valid nodes.

Reply:
in what case would the mons be not in a valid node? Before a mon is placed into a node, it calls `getAvailableMonNodes()`, which should return a valid node and the mon gets placed there. So shouldnt the mons always be in a valid nodes to begin with?

Reply:
@kokhang When a) node's label are modified it could happen that a mon isn't "allowed" to be on the node anymore through the `placement` spec and b) the `placement` for mons gets updated.

Reply:
ahh ok.. that could happen when the taint is set to something other than`NoExecute`.